File: neoforge/src/main/java/com/eliasnvx/krylix/forge/PlayerKillStatsData.java
package com.eliasnvx.krylix.forge;

import com.eliasnvx.krylix.Krylix;
import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.resources.Identifier;
import net.minecraft.server.MinecraftServer;
import net.minecraft.util.datafix.DataFixTypes;
import net.minecraft.world.level.saveddata.SavedData;
import net.minecraft.world.level.saveddata.SavedDataType;

import java.util.HashMap;
import java.util.Map;

public class PlayerKillStatsData extends SavedData {
    public final Map<String, PlayerStat> stats = new HashMap<>();

    public static final Codec<PlayerKillStatsData> CODEC = RecordCodecBuilder.create(instance ->
        instance.group(
            Codec.unboundedMap(Codec.STRING, PlayerStat.CODEC).optionalFieldOf("stats", Map.of()).forGetter(d -> d.stats)
        ).apply(instance, map -> {
            PlayerKillStatsData data = new PlayerKillStatsData();
            data.stats.putAll(map);
            return data;
        })
    );

    public static final SavedDataType<PlayerKillStatsData> TYPE = new SavedDataType<>(
        Identifier.fromNamespaceAndPath(Krylix.MOD_ID, "player_stats"),
        PlayerKillStatsData::new,
        CODEC,
        DataFixTypes.LEVEL
    );

    public static PlayerKillStatsData get(MinecraftServer server) {
        return server.overworld().getDataStorage().computeIfAbsent(TYPE);
    }

    public void recordKill(String killerUuid, String killerName) {
        PlayerStat stat = stats.computeIfAbsent(killerUuid, k -> new PlayerStat());
        stat.kills++;
        stat.lastName = killerName;
        setDirty();
    }

    public void recordDeath(String victimUuid, String victimName) {
        PlayerStat stat = stats.computeIfAbsent(victimUuid, k -> new PlayerStat());
        stat.deaths++;
        stat.lastName = victimName;
        setDirty();
    }

    public void recordMobKill(String killerUuid, String killerName) {
        PlayerStat stat = stats.computeIfAbsent(killerUuid, k -> new PlayerStat());
        stat.mobKills++;
        stat.lastName = killerName;
        setDirty();
    }
}

File: forge/src/main/kotlin/com/eliasnvx/krylix/forge/PlayerKillStatsData.kt
package com.eliasnvx.krylix.forge

import net.minecraft.nbt.CompoundTag
import net.minecraft.server.MinecraftServer
import net.minecraft.world.level.saveddata.SavedData

data class PlayerStat(
    var kills: Int = 0,
    var deaths: Int = 0,
    var mobKills: Int = 0,
    var lastName: String = "",
)

/**
 * Рейтинг игроков по PvP-убийствам и личным убийствам мобов за весь мир (per-world, аналогично
 * MobKillStatsData — тот считает "какого моба убивали чаще всего", а этот "кто из игроков убил
 * больше мобов"). Ключ — UUID игрока строкой (стабилен в отличие от ника).
 */
class PlayerKillStatsData : SavedData() {
    val stats: MutableMap<String, PlayerStat> = mutableMapOf()

    fun recordKill(killerUuid: String, killerName: String) {
        val stat = stats.getOrPut(killerUuid) { PlayerStat() }
        stat.kills++
        stat.lastName = killerName
        setDirty()
    }

    fun recordDeath(victimUuid: String, victimName: String) {
        val stat = stats.getOrPut(victimUuid) { PlayerStat() }
        stat.deaths++
        stat.lastName = victimName
        setDirty()
    }

    fun recordMobKill(killerUuid: String, killerName: String) {
        val stat = stats.getOrPut(killerUuid) { PlayerStat() }
        stat.mobKills++
        stat.lastName = killerName
        setDirty()
    }

    override fun save(tag: CompoundTag): CompoundTag {
        val statsTag = CompoundTag()
        stats.forEach { (uuid, stat) ->
            val entryTag = CompoundTag()
            entryTag.putInt("kills", stat.kills)
            entryTag.putInt("deaths", stat.deaths)
            entryTag.putInt("mobKills", stat.mobKills)
            entryTag.putString("name", stat.lastName)
            statsTag.put(uuid, entryTag)
        }
        tag.put("stats", statsTag)
        return tag
    }

    companion object {
        private const val ID = "krylix_player_stats"

        private fun load(tag: CompoundTag): PlayerKillStatsData {
            val data = PlayerKillStatsData()
            val statsTag = tag.getCompound("stats")
            for (uuid in statsTag.allKeys) {
                val entryTag = statsTag.getCompound(uuid)
                data.stats[uuid] = PlayerStat(
                    kills = entryTag.getInt("kills"),
                    deaths = entryTag.getInt("deaths"),
                    mobKills = entryTag.getInt("mobKills"),
                    lastName = entryTag.getString("name"),
                )
            }
            return data
        }

        fun get(server: MinecraftServer): PlayerKillStatsData =
            server.overworld().dataStorage.computeIfAbsent(::load, ::PlayerKillStatsData, ID)
    }
}
